MBS-13711 (II): Warn about capitalized short words in titles

derat opened the pull request:
# MBS-13711

Make the tracklist editor display a warning if an apparently-English title capitalizes a short word that should be lowercased. "and", "of", "or", "to", and "the" are checked. A confirmation checkbox is displayed for beginner editors.

# Problem

Beginner editors often import releases from Discogs (where the [First Letter Of Each Word Is Capitalized](https://support.discogs.com/hc/en-us/articles/360005006334-Database-Guidelines-1-General-Rules#Capitalization_And_Grammar)) without updating titles to follow [MusicBrainz's language-specific guidelines](https://musicbrainz.org/doc/Style#Language_specific_guidelines). This creates cleanup work for other editors.

# Solution

Use a similar approach to #3504 to display a warning pointing to the English style guidelines if any of the above words appear in the middle of a track title with their first word capitalized.

![Screenshot 2025-03-27 07 31 37](https://github.com/user-attachments/assets/8622e107-88b6-453f-ad3f-efb499b710e6)
_Verdana has ugly curly/angled double quotes. :-(_

# Testing

Entered miscapitalized titles as a beginner and non-beginner editor and with unset/English/other languages and Japan/non-Japan release events.

derat commented on root/static/scripts/release-editor/fields.js:
```diff
@@ -387,6 +398,11 @@ class Track {
   }
 
+  hasMiscapitalizedTitle() {
+    return this.medium.release.shouldUseEnglishCapitalization() &&
```
I'm wondering if I should try to find some way to also exclude releases and/or tracks credited to artists with Japanese areas, since this may annoy people who edit a lot of English-language Japanese releases otherwise. It might be a bit tricky, since it looked to me like artist areas aren't currently being passed to the release editor (I didn't dig deeper to see if they're being loaded from the DB, but I suspect they may not be).

On the other hand, this change just displays the yellow warning box without the checkbox for non-beginners, so maybe it's not a big deal. `¯\_(ツ)_/¯`

reosarevok replied:
I'm fairly sure they will survive. They already get an exception, they can live with ignoring a yellow box.

derat commented on root/static/scripts/release-editor/fields.js:
```diff
@@ -387,6 +398,11 @@ class Track {
 
+  hasMiscapitalizedTitle() {
+    return this.medium.release.shouldUseEnglishCapitalization() &&
+      miscapitalizedEnglishRegex.test(this.name());
```
One potential problem that I noticed is that this function (and also `hasFeatInTitle` and `hasCapitalizedETI`) run for every keystroke while the user is typing into a title input. I think that it's happening due to `Track`'s `name` member being updated on `input` rather than `change`.

This could have performance implications, and it might also be annoying from a UX perspective (although I think that all three checks are unlikely to be triggered incorrectly mid-title). I tried working around it by wrapping these functions in `debounceComputed`, but that has the downside of making them to be slower to give feedback, and it also means that the warnings take half a second to disappear after the user clicks "Guess feat." or "Guess case" to fix the titles.

derat replied:
I think I'd be happier with debouncing if there were some way to trigger evaluation immediately after "Guess case" et al is clicked, but it wasn't obvious to me from [the Knockout docs](https://knockoutjs.com/documentation/rateLimit-observable.html) whether that's possible.

I was worried at first about warnings showing up for e.g. "The" while the user is manually typing "There", but I think that that's unlikely due to the regexes looking for trailing whitespace after the error.

mwiencek replied:
What about specifying a shorter debounce interval (second argument to `debounceComputed`)? It defaults to 500ms, but 250ms is probably enough to avoid updates between keystrokes while reducing the update lag after hitting the guess-case button too.

derat replied:
250 ms feels pretty good to me; thanks for the suggestion. I've also updated the methods for feat. in titles and capitalized ETI to be debounced.

Aerozol replied:
When this runs and it triggers the note, I guess it will shift the page down a bit? It could be annoying in terms of UX, including with the delay.
But my overall impression is that not having to fix endless #^%ing Discogs imports is more than worth it :)
I think this would best be tweaked after feedback in beta - though beginner users might not give feedback, which is a different issue.

derat replied:
@Aerozol, yeah, there's unfortunately a visible shift. I couldn't think of any way to avoid it apart from reserving empty space at the top of the page (ugh) or displaying the message somewhere else (which may require a fair amount of re-engineering; I just piggybacked on the current warning-displaying logic for this).

I'm hoping it won't be a huge deal, since (barring false positives) it'll be displayed when people aren't following the guidelines, plus most editors seem to be seeding releases now so the warning should already be there when they switch to the tracklist tab. The shift might be weird if someone is manually typing titles and has entered enough tracks to scroll the screen down so they can't see the warning, though...

derat commented:
@reosarevok, @mwiencek, and @Aerozol, mind taking a look at this one too? I left a few comments about open questions that I have.

reosarevok commented on root/static/scripts/release-editor/fields.js:
```diff
@@ -59,6 +59,13 @@ const bracketRegex = new RegExp(
+ * ../guess-case/modes.js.
+ */
+const miscapitalizedEnglishRegex =
+  /[a-z]\s+(?:And|In|Of|Or|The|To)\s+[a-zA-Z]/;
```
You could consider omitting "in" as well then since it's probably a lot less common than something like "and" or "of" or "the" anyway, so in many cases if "in" is there wrongly I would expect one of the others to maybe be too.

I don't think even native English speakers often clock the difference between part-of-verb and preposition, tbh.

reosarevok replied:
I was testing this and was confused about why it wouldn't trigger, then I noticed it's because we expect `[a-z]` before the space and I had a capital letter. Is there any real reason for this? Admittedly it's probably not super common, but this doesn't trigger on something like "Available On CD And Vinyl"

derat replied:
Sure, I've removed "In" here and from the explanation in the template.

I've updated the regex to support earlier capital letters. I think my initial intent was to not match all-caps titles, but that would probably only be an issue if I included "A" in the group, and I didn't do that due to the risk of false positives.

Requiring leading and trailing letters felt a bit safer, but I think you're right that it'd probably be safe for the "before" character class to be something like `[a-zA-Z0-9,]`. Including dashes and semicolons seems like it might be riskier since I could imagine those being used to separate sentences or phrases that should be independently capitalized. I didn't want to put numbers in the "after" class since it could match properly-capitalized final words in numbered song titles, e.g. imagine a pair of songs named "Killing in the Name Of" and "Killing in the Name Of 2" (but let me know if "of" should actually be lowercase in the second one). :-)

reosarevok commented on root/release/edit/tracklist.tt:
```diff
@@ -456,6 +456,25 @@
+        [%- warning_icon %]
+        <p>
+          <strong>[% l('Warning:') | html_entity %]</strong>
+          [% l('This medium may have incorrectly-capitalized {english|English track titles}. Short articles, conjunctions, and prepositions like “and”, “in”, “of”, “or”, “to”, and “the” should be typically be lowercased.', { english => { href => doc_link('Style/Language/English'), target => '_blank' } }) %]
```
"incorrectly-capitalized" might be correct with the hyphen too, but it makes it harder to read IMO. I'd just use a space.

"should be typically be lowercased" has a be too many.

Would it be better to have the link to the guidelines in "should typically be lowercased"?

reosarevok commented on root/static/scripts/release-editor/fields.js:
```diff
@@ -88,6 +95,10 @@ class Track {
       capitalizedETIRegex.test(data.name);
+    this.hasMiscapitalizedOrigTitle =
+      isDatabaseRowId(this.id) &&
+      medium.release.shouldUseEnglishCapitalization() &&
```
Should this also be used on `hasCapitalizedETIInOrigTitle` above?

reosarevok reviewed (commented):
Tested this as a beginner and a non-beginner it works nicely in most cases, left some comments though with cases where it could be improved IMO.

derat commented:
Looks like the `js-perl-and-pgtap` tests failed again:

```
  ...
  switching to the tracklist tab opens the add-medium dialog if thereâs only one empty medium
    ✓ add-medium dialog is open after switching to the tracklist tab
    ✓ add-medium dialog is closed after switching back to the information tab
    Script error.
    File name: 
    Line number: 0
ERROR: Test timed out
```

I'm able to repro the failure locally; it's due to this:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'release')
    at Track.hasMiscapitalizedTitle (fields.js:402:24)
    at fields.js:567:40
    at Array.some (<anonymous>)
    at fields.js:567:28
```

Here's the code:

```js
 401   hasMiscapitalizedTitle() {          
 402     return this.medium.release.shouldUseEnglishCapitalization() &&
 403       miscapitalizedEnglishRegex.test(this.name());
 404   }
```

Using optional chaining (`?.`) here (and when initializing `hasMiscapitalizedOrigTitle` in the c'tor) avoids this, but is it expected that the medium can be undefined when creating a track? Maybe so, since it's also checked later in the c'tor:

```js
 121     var release = medium && medium.release;
 122 
 123     if (release &&
 124         !data.artistCredit &&
 125         !hasVariousArtists(release.artistCredit.peek())) {
 126       data.artistCredit = release.artistCredit.peek();
 127     }
```

mwiencek commented:
> is it expected that the medium can be undefined when creating a track?

Ah, I think that one line in the dialogs test is the only place we don't pass a medium. Just fixed that in https://github.com/metabrainz/musicbrainz-server/commit/b44437023a61bef3a561d4a5ea6afa5bfc00a95a if you'd like to rebase on master and remove the optional chaining.

derat commented:
> > is it expected that the medium can be undefined when creating a track?
> 
> Ah, I think that one line in the dialogs test is the only place we don't pass a medium. Just fixed that in [b444370](https://github.com/metabrainz/musicbrainz-server/commit/b44437023a61bef3a561d4a5ea6afa5bfc00a95a) if you'd like to rebase on master and remove the optional chaining.

Thanks! I've rebased and removed it now.

---

Is there some easy way to see which tests failed from a page like https://app.circleci.com/pipelines/github/metabrainz/musicbrainz-server/13016/workflows/ac288826-5824-432f-837b-3a18e6edf115/jobs/21348? I see `Failed 85/427 subtests`, but I'm not sure what that's referring to -- were the failures in subtests of `t::MusicBrainz::Server::Edit::Relationship::AddLinkType`?

The Selenium failure at https://ci.metabrainz.org/job/musicbrainz-server/12006/console seems like it might be unrelated to this change, but I think it's in a test I added a while back (`MBS-2604: Infer link direction between people and groups`). :-P

mwiencek commented:
> Is there some easy way to see which tests failed from a page like https://app.circleci.com/pipelines/github/metabrainz/musicbrainz-server/13016/workflows/ac288826-5824-432f-837b-3a18e6edf115/jobs/21348? I see `Failed 85/427 subtests`, but I'm not sure what that's referring to -- were the failures in subtests of `t::MusicBrainz::Server::Edit::Relationship::AddLinkType`?

Hmm, normally ctrl+f for "not ok" will find them, but in this case it looks like the tests process just completely aborts at "Can create relationship types with documentation," which I'm not sure of the cause of...if it happens again on a rerun I can take another look locally.

Let's run the Selenium tests again too. 😅

@brainzbot test this please

derat commented:
@mwiencek, thanks for the test fixes!

reosarevok reviewed (approved):
This seems good and ready to merge, but I have one last minute small suggestion! 

I might at some point soon shamelessly reuse most of this code to warn about "In B Major" style classical titles...

reosarevok commented on root/static/scripts/release-editor/fields.js:
```diff
@@ -59,6 +59,15 @@ const bracketRegex = new RegExp(
+ * ../guess-case/modes.js.
+ */
+const miscapitalizedEnglishRegex =
+  /[a-zA-Z0-9,]\s+(?:And|Of|Or|The|To)\s+[a-zA-Z]/;
```
I just re-read your answer from a few days ago saying that

> I didn't want to put numbers in the "after" class since it could match properly-capitalized final words in numbered song titles, e.g. imagine a pair of songs named "Killing in the Name Of" and "Killing in the Name Of 2" (but let me know if "of" should actually be lowercase in the second one). :-)

It should be uppercase, but I kinda feel this is uncommon enough it might not be too relevant anyway (how often does this happen?). But I guess my example is also not super common so I can live without if you want to play it very safe :)

derat replied:
I was worried that this could introduce false positives in some cases:

> I didn't want to put numbers in the "after" class since it could match properly-capitalized final words in numbered song titles, e.g. imagine a pair of songs named "Killing in the Name Of" and "Killing in the Name Of 2" (but let me know if "of" should actually be lowercase in the second one). :-)

If that isn't a concern, adding digits to the class sounds fine to me, though! It might be safest if @mwiencek makes the change since I think he's added some more commits to this PR and I don't want to mess things up.

reosarevok replied:
Merging for now, we can always add numbers later if we see it's an issue.
